L1Trigger/TrackerTFP/src/KalmanFilter.cc

```diff
@@ -0,0 +1,398 @@
+    K10_(&kalmanFilterFormats_->format(VariableKF::K10)),
+    K21_(&kalmanFilterFormats_->format(VariableKF::K21)),
+    K31_(&kalmanFilterFormats_->format(VariableKF::K31)),
+    R00_(&kalmanFilterFormats_->format(VariableKF::R00)),
+    R11_(&kalmanFilterFormats_->format(VariableKF::R11)),
+    R00Rough_(&kalmanFilterFormats_->format(VariableKF::R00Rough)),
+    R11Rough_(&kalmanFilterFormats_->format(VariableKF::R11Rough)),
+    invR00Approx_(&kalmanFilterFormats_->format(VariableKF::invR00Approx)),
+    invR11Approx_(&kalmanFilterFormats_->format(VariableKF::invR11Approx)),
+    invR00Cor_(&kalmanFilterFormats_->format(VariableKF::invR00Cor)),
+    invR11Cor_(&kalmanFilterFormats_->format(VariableKF::invR11Cor)),
+    invR00_(&kalmanFilterFormats_->format(VariableKF::invR00)),
+    invR11_(&kalmanFilterFormats_->format(VariableKF::invR11)),
+    C00_(&kalmanFilterFormats_->format(VariableKF::C00)),
+    C01_(&kalmanFilterFormats_->format(VariableKF::C01)),
+    C11_(&kalmanFilterFormats_->format(VariableKF::C11)),
+    C22_(&kalmanFilterFormats_->format(VariableKF::C22)),
+    C23_(&kalmanFilterFormats_->format(VariableKF::C23)),
+    C33_(&kalmanFilterFormats_->format(VariableKF::C33)) {}
+
+  // read in and organize input tracks and stubs
+  void KalmanFilter::consume(const StreamsTrack& streamsTrack, const TTDTC::Streams& streamsStub) {
+    auto valid = [](const auto& frame){ return frame.first.isNonnull(); };
+    auto acc = [](int& sum, const auto& frame){ return sum += (frame.first.isNonnull() ? 1 : 0); };
+    int nTracks(0);
+    int nStubs(0);
+    const int offset = region_ * dataFormats_->numChannel(Process::kf);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::kf); channel++) {
+      const int channelTrack = offset + channel;
+      const StreamTrack& streamTracks = streamsTrack[channelTrack];
+      nTracks += accumulate(streamTracks.begin(), streamTracks.end(), 0, acc);
+      for (int layer = 0; layer < setup_->numLayers(); layer++) {
+        const int channelStub = channelTrack * setup_->numLayers() + layer;
+        const TTDTC::Stream& streamStubs = streamsStub[channelStub];
+        nStubs += accumulate(streamStubs.begin(), streamStubs.end(), 0, acc);
+      }
+    }
+    tracks_.reserve(nTracks);
+    stubs_.reserve(nStubs);
+    for (int channel = 0; channel < dataFormats_->numChannel(Process::kf); channel++) {
+      const int channelTrack = offset + channel;
+      const StreamTrack& streamTracks = streamsTrack[channelTrack];
+      vector<TrackKFin*>& tracks = input_[channel];
+      tracks.reserve(streamTracks.size());
+      for (int frame = 0; frame < (int)streamTracks.size(); frame++) {
+        const FrameTrack& frameTrack = streamTracks[frame];
+        if (frameTrack.first.isNull())
+          continue;
+        const auto end = find_if(next(streamTracks.begin(), frame + 1), streamTracks.end(), valid);
+        const int size = distance(next(streamTracks.begin(), frame), end);
```

Comment:
Add comment "No. of frames before next track indicates gives max. no. of stubs this track had in any layer"

Reply:
done